go-controller/pkg/ovn/policy.go

```diff
@@ -1703,12 +1758,36 @@ func (oc *Controller) addPeerNamespaceAndPodHandler(
 	return nil
 }
 
-func (oc *Controller) handlePeerNamespaceSelectorOnUpdate(np *networkPolicy, gp *gressPolicy, doUpdate func() bool) error {
+func (oc *Controller) handlePeerNamespaceSelectorAdd(np *networkPolicy, gp *gressPolicy, obj interface{}) error {
+	namespace := obj.(*kapi.Namespace)
+	np.Lock()
+	updated := gp.addNamespaceAddressSet(namespace.Name)
+	np.Unlock()
+	// unlock networkPolicy, before calling peerNamespaceUpdate
+	if updated {
+		return oc.peerNamespaceUpdate(np, gp)
+	}
+	return nil
+}
+
+func (oc *Controller) handlePeerNamespaceSelectorDel(np *networkPolicy, gp *gressPolicy, obj interface{}) error {
+	namespace := obj.(*kapi.Namespace)
+	np.Lock()
+	updated := gp.delNamespaceAddressSet(namespace.Name)
+	np.Unlock()
+	// unlock networkPolicy, before calling peerNamespaceUpdate
+	if updated {
+		return oc.peerNamespaceUpdate(np, gp)
+	}
+	return nil
+}
+
+func (oc *Controller) peerNamespaceUpdate(np *networkPolicy, gp *gressPolicy) error {
 	aclLoggingLevels := oc.GetNamespaceACLLogging(np.namespace)
 	np.Lock()
 	defer np.Unlock()
 	// This needs to be a write lock because there's no locking around 'gress policies
-	if !np.deleted && doUpdate() {
+	if !np.deleted {
```

Comment:
So the previous implementation actually wasn't correct, because it didn't hold namespace lock while updating acls with namespace logleves, so if namespace update event happened between aclloglevel was saved and db transaction, and acls were updated with the new loglevel, this handler will overwrite it with the stale saved loglevel.
To lock network policy only once, we will have to lock namespace first, then check if update is required and update acls if so.
To save namespace locks, I check first if update is required, and lock namespace only in that case. That is safe because `peerNamespaceUpdate` just takes `gressPolicy`, so as you said, it will update ACLs to the latest state, and the following calls will be noop. Also we need to check that `buildLocalPodACLs` doesn't read something that other thread may be writing at the same time. To verify that we check `gressPolicy` object, the only fields that are updated after creation is peerAddressSet, peerV4AddressSets, peerV6AddressSets. Last two are safe for concurrent use, and the first one is not used in `buildLocalPodACLs`.
I guess I need to add some comments around that :D